File: skellig-test-step-processing-rmq/src/main/java/org/skellig/teststep/processor/rmq/RmqDetailsConfigReader.java
package org.skellig.teststep.processor.rmq;

import com.typesafe.config.Config;
import org.skellig.teststep.processor.rmq.model.RmqDetails;
import org.skellig.teststep.processor.rmq.model.RmqExchangeDetails;
import org.skellig.teststep.processor.rmq.model.RmqHostDetails;
import org.skellig.teststep.processor.rmq.model.RmqQueueDetails;

import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;


class RmqDetailsConfigReader {

    private static final String RMQ_CONFIG_KEYWORD = "rmq";

    Collection<RmqDetails> read(Config config) {
        Objects.requireNonNull(config, "RMQ config cannot be null");

        Collection<RmqDetails> rmqDetails = Collections.emptyList();
        if (config.hasPath(RMQ_CONFIG_KEYWORD)) {
            List<Map> anyRefList = (List<Map>) config.getAnyRefList(RMQ_CONFIG_KEYWORD);
            rmqDetails = anyRefList.stream()
                    .map(this::createRmqDetails)
                    .flatMap(Collection::stream)
                    .collect(Collectors.toList());
        }
        return rmqDetails;
    }

    private Collection<RmqDetails> createRmqDetails(Map rawRmqDetails) {
        RmqHostDetails hostDetails = createRmqHostDetails(rawRmqDetails);
        List<Map> queues = (List<Map>) rawRmqDetails.get("queues");
        Map<String, RmqExchangeDetails> exchanges = createExchanges(rawRmqDetails);

        return queues.stream()
                .map(queue -> createQueueDetails(queue, exchanges, hostDetails))
                .collect(Collectors.toList());
    }

    private RmqHostDetails createRmqHostDetails(Map rawRmqDetails) {
        String host = (String) rawRmqDetails.get("host");
        int port = (int) rawRmqDetails.get("port");
        String user = (String) rawRmqDetails.get("username");
        String password = (String) rawRmqDetails.get("password");
        return new RmqHostDetails(host, port, user, password);
    }

    private Map<String, RmqExchangeDetails> createExchanges(Map rawExchangesDetails) {
        List<Map> exchanges = (List<Map>) rawExchangesDetails.get("exchanges");

        Objects.requireNonNull(exchanges, "No exchanges were declared for RMQ");

        return exchanges.stream()
                .map(this::createExchange)
                .collect(Collectors.toMap(RmqExchangeDetails::getName, e -> e));
    }

    private RmqExchangeDetails createExchange(Map rawExchangeDetails) {
        String name = (String) rawExchangeDetails.get("name");
        String type = (String) rawExchangeDetails.get("type");

        Objects.requireNonNull(name, "Name was not declared for RMQ Exchange");

        return new RmqExchangeDetails.Builder()
                .withName(name)
                .withType(type)
                .withDurable(extractIsDurable(rawExchangeDetails))
                .withAutoDelete(extractIsAutoDelete(rawExchangeDetails))
                .withCreateIfNew(extractCreateIfNew(rawExchangeDetails))
                .withParameters(extractParameters(rawExchangeDetails))
                .build();
    }

    private RmqDetails createQueueDetails(Map item, Map<String, RmqExchangeDetails> exchanges,
                                          RmqHostDetails hostDetails) {
        String channelId = (String) item.get("channelId");
        String name = (String) item.get("name");
        String exchange = (String) item.get("exchange");

        Objects.requireNonNull(channelId, "Channel ID was not declared for RMQ details. " +
                "It can be any unique name which you would use in tests as a reference");
        Objects.requireNonNull(name, "Queue name was not declared for RMQ details");
        Objects.requireNonNull(exchange, "Exchange name was not declared for RMQ details");
        Objects.requireNonNull(exchanges.get(exchange), String.format("No exchange name '%s' was declared", exchange));

        RmqQueueDetails queue = new RmqQueueDetails.Builder()
                .withName(name)
                .withRoutingKey(extractRoutingKey(item))
                .withDurable(extractIsDurable(item))
                .withAutoDelete(extractIsAutoDelete(item))
                .withCreateIfNew(extractCreateIfNew(item))
                .withExclusive(extractIsExclusive(item))
                .withParameters(extractParameters(item))
                .build();

        return new RmqDetails.Builder()
                .withChannelId(channelId)
                .withHostDetails(hostDetails)
                .withQueue(queue)
                .withExchange(exchanges.get(exchange))
                .build();
    }

    private String extractRoutingKey(Map rawQueueDetails) {
        String routingKey = (String) rawQueueDetails.get("routingKey");
        return routingKey == null ? "#" : routingKey;
    }

    private boolean extractIsExclusive(Map rawQueueDetails) {
        Boolean isExclusive = (Boolean) rawQueueDetails.get("exclusive");
        return isExclusive == null ? false : isExclusive;
    }

    private boolean extractIsAutoDelete(Map rawQueueDetails) {
        Boolean autoDelete = (Boolean) rawQueueDetails.get("autoDelete");
        return autoDelete == null ? false : autoDelete;
    }

    private boolean extractCreateIfNew(Map rawQueueDetails) {
        Boolean create = (Boolean) rawQueueDetails.get("create");
        return create == null ? false : create;
    }

    private boolean extractIsDurable(Map rawQueueDetails) {
        Boolean isDurable = (Boolean) rawQueueDetails.get("durable");
        return isDurable == null ? false : isDurable;
    }

    private Map<String, Object> extractParameters(Map rawQueueDetails) {
        Map<String, Object> parameters = (Map<String, Object>) rawQueueDetails.get("parameters");
        return parameters == null ? Collections.emptyMap() : parameters;
    }
}

File: skellig-test-step-processing-rmq/src/main/kotlin/org/skellig/teststep/processor/rmq/RmqDetailsConfigReader.kt
package org.skellig.teststep.processor.rmq

import com.typesafe.config.Config
import org.skellig.teststep.processing.util.logger
import org.skellig.teststep.processor.rmq.model.RmqDetails
import org.skellig.teststep.processor.rmq.model.RmqExchangeDetails
import org.skellig.teststep.processor.rmq.model.RmqHostDetails
import org.skellig.teststep.processor.rmq.model.RmqQueueDetails
import java.util.*

internal class RmqDetailsConfigReader {

    companion object {
        private const val RMQ_CONFIG_KEYWORD = "rmq.hosts"
    }

    private val log = logger<RmqDetailsConfigReader>()

    fun read(config: Config?): Collection<RmqDetails> {
        Objects.requireNonNull(config, "RMQ config cannot be null")

        var rmqDetails: Collection<RmqDetails> = emptyList()
        if (config!!.hasPath(RMQ_CONFIG_KEYWORD)) {
            log.info("RMQ configuration found in the Config file. Start to register its queue")
            val anyRefList = config.getAnyRefList(RMQ_CONFIG_KEYWORD) as List<*>
            rmqDetails = anyRefList
                .map { createRmqDetails(it) }
                .flatten()
                .toList()
        }
        return rmqDetails
    }

    private fun createRmqDetails(rawRmqDetails: Any?): Collection<RmqDetails> {
        return (rawRmqDetails as Map<*, *>?)?.let {
            val hostDetails = createRmqHostDetails(it)
            val queues = it["queues"] as List<*>?
            val exchanges = createExchanges(it)
            queues!!
                .mapNotNull { createQueueDetails(it, exchanges, hostDetails) }
                .toList()
        } ?: emptyList()
    }

    private fun createRmqHostDetails(rawRmqDetails: Map<*, *>): RmqHostDetails {
        val host = rawRmqDetails["host"] as String
        val port = rawRmqDetails["port"] as Int
        val user = rawRmqDetails["username"] as String?
        val password = rawRmqDetails["password"] as String?

        return RmqHostDetails(host, port, user, password)
    }

    private fun createExchanges(rawExchangesDetails: Map<*, *>): Map<String, RmqExchangeDetails> {
        val exchanges = (rawExchangesDetails["exchanges"] ?: error("No exchanges were declared for RMQ")) as List<*>?

        return exchanges!!
            .mapNotNull { createExchange(it) }.associateBy { it.name }
    }

    private fun createExchange(rawExchangeDetails: Any?): RmqExchangeDetails? {
        return (rawExchangeDetails as Map<*, *>?)?.let {
            val name = (it["name"] ?: error("Name was not declared for RMQ Exchange")) as String?
            val type = it["type"] as String?

            return RmqExchangeDetails.Builder()
                .name(name)
                .type(type)
                .durable(extractIsDurable(it))
                .autoDelete(extractIsAutoDelete(it))
                .createIfNew(extractCreateIfNew(it))
                .parameters(extractParameters(it))
                .build()
        }
    }

    private fun createQueueDetails(
        item: Any?,
        exchanges: Map<String, RmqExchangeDetails>,
        hostDetails: RmqHostDetails
    ): RmqDetails? {
        return (item as Map<*, *>?)?.let {
            val id = it["id"] as String?
            val name = (it["name"] ?: error("Queue name was not declared for RMQ details")) as String
            val exchange = it["exchange"] ?: error("Exchange name was not declared for RMQ details")

            Objects.requireNonNull(exchanges[exchange], String.format("No exchange name '%s' was declared", exchange))

            val queue = RmqQueueDetails.Builder()
                .id(id)
                .name(name)
                .routingKey(extractRoutingKey(it))
                .durable(extractIsDurable(it))
                .autoDelete(extractIsAutoDelete(it))
                .createIfNew(extractCreateIfNew(it))
                .exclusive(extractIsExclusive(it))
                .parameters(extractParameters(it))
                .build()

            RmqDetails.Builder()
                .hostDetails(hostDetails)
                .queue(queue)
                .exchange(exchanges[exchange])
                .build()
        }
    }

    private fun extractRoutingKey(rawQueueDetails: Map<*, *>): String {
        val routingKey = rawQueueDetails["routingKey"] as String?
        return routingKey ?: "#"
    }

    private fun extractIsExclusive(rawQueueDetails: Map<*, *>): Boolean {
        val isExclusive = rawQueueDetails["exclusive"] as Boolean?
        return isExclusive ?: false
    }

    private fun extractIsAutoDelete(rawQueueDetails: Map<*, *>): Boolean {
        val autoDelete = rawQueueDetails["autoDelete"] as Boolean?
        return autoDelete ?: false
    }

    private fun extractCreateIfNew(rawQueueDetails: Map<*, *>): Boolean {
        val create = rawQueueDetails["create"] as Boolean?
        return create ?: false
    }

    private fun extractIsDurable(rawQueueDetails: Map<*, *>): Boolean {
        val isDurable = rawQueueDetails["durable"] as Boolean?
        return isDurable ?: false
    }

    private fun extractParameters(rawQueueDetails: Map<*, *>): MutableMap<String, Any>? {
        return rawQueueDetails["parameters"] as MutableMap<String, Any>?
    }
}